File: backend/app/schemas/realtime.py
from enum import StrEnum
from typing import Literal
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field, field_validator


class RealtimeEventType(StrEnum):
    GROCERY_ITEM_ADDED = "grocery.item_added"
    GROCERY_ITEM_EDITED = "grocery.item_edited"
    GROCERY_ITEM_COMPLETED = "grocery.item_completed"
    GROCERY_ITEM_REOPENED = "grocery.item_reopened"
    GROCERY_ITEM_DELETED = "grocery.item_deleted"


class GroceryItemRealtimePayload(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    shopping_session_id: UUID
    grocery_item_id: UUID
    actor_user_id: UUID | None
    item_name: str = Field(min_length=1, max_length=160)
    sequence_number: int = Field(gt=0)

    @field_validator("item_name", mode="before")
    @classmethod
    def normalize_item_name(cls, value: object) -> object:
        if not isinstance(value, str):
            return value

        normalized = " ".join(value.split())
        if not normalized:
            raise ValueError("Grocery item name cannot be blank.")
        return normalized


class RealtimeEventEnvelope(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    schema_version: Literal[1] = 1
    event_id: UUID
    event_type: RealtimeEventType
    household_id: UUID
    occurred_at: AwareDatetime
    payload: GroceryItemRealtimePayload

File: mobile/src/types/realtime.ts
import { z } from "zod";

export const realtimeEventTypes = [
  "grocery.item_added",
  "grocery.item_edited",
  "grocery.item_completed",
  "grocery.item_reopened",
  "grocery.item_deleted",
] as const;

export const groceryItemRealtimePayloadSchema = z.strictObject({
  shopping_session_id: z.uuid(),
  grocery_item_id: z.uuid(),
  actor_user_id: z.uuid().nullable(),
  item_name: z
    .string()
    .min(1)
    .max(160)
    .refine((value) => value.trim().length > 0),
  sequence_number: z.number().int().positive(),
});

export const realtimeEventSchema = z.strictObject({
  schema_version: z.literal(1),
  event_id: z.uuid(),
  event_type: z.enum(realtimeEventTypes),
  household_id: z.uuid(),
  occurred_at: z.iso.datetime({ offset: true }),
  payload: groceryItemRealtimePayloadSchema,
});

export type RealtimeEvent = z.infer<typeof realtimeEventSchema>;
export type RealtimeEventType = RealtimeEvent["event_type"];
